tensorflow/compiler/jit/deadness_analysis.cc

```diff
@@ -636,6 +643,35 @@ Predicate* PredicateFactory::MakeAndOrImpl(
     negated_ops.insert(negated_op);
   }
 
+  // Simplify {S,&,X} & ~X & ... => S & ...
+  if (is_and) {
+    absl::flat_hash_set<Predicate*> to_remove;
+    std::vector<Predicate*> to_add;
+    for (Predicate* op : simplified_ops) {
+      if (op->kind() == Predicate::Kind::kAndRecurrence) {
+        auto* and_rec = static_cast<AndRecurrencePredicate*>(op);
+        if (negated_ops.contains(and_rec->step())) {
+          // Remove and_rec and ~X and insert S.  Note that checking the
+          // existence of ~X through negated_ops is sufficient since it makes
+          // sure the predicate is in the input operands.  It does not need to
+          // be in simplified_ops if it was already cancelled out.
+          to_remove.insert(and_rec);
+          to_remove.insert(MakeNotPredicate(and_rec->step()));
+          to_add.push_back(and_rec->start());
+        }
+      }
+    }
+    auto it = simplified_ops.begin();
+    while (it != simplified_ops.end()) {
+      if (to_remove.contains(*it)) {
```

Comment:
I don't get this. `simplified_ops` seen in `simplified_ops`?

Reply:
If you meant to check the op seen in `negation_ops` has the correspondents in `simplified_ops`, it should not be necessary because it is logically correct the op is not in `simplified_ops` (after it is cancelled out.)

Reply:
Argh, sorry for the weird comment.  I meant we should `DCHECK` that all elements in `to_remove` were seen in `simplified_ops`.

Reply:
I'll send in a PR once this is checked in to implement this.